geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/utils/SortingSimpleFeatureIterator.scala

```diff
@@ -0,0 +1,289 @@
+import org.locationtech.geomesa.features.SimpleFeatureSerializer
+import org.locationtech.geomesa.features.kryo.KryoFeatureSerializer
+import org.locationtech.geomesa.index.conf.QueryProperties
+import org.locationtech.geomesa.utils.collection.CloseableIterator
+import org.locationtech.geomesa.utils.geotools.SimpleFeatureOrdering
+import org.locationtech.geomesa.utils.index.ByteArrays
+import org.locationtech.geomesa.utils.io.{CloseQuietly, Sizable, WithClose}
+import org.opengis.feature.simple.SimpleFeature
+
+import scala.collection.mutable.ArrayBuffer
+import scala.util.Random
+
+/**
+  * In memory sorting of simple features, with optional spill to disk
+  *
+  * @param features unsorted feature iterator
+  * @param sortBy attributes to sort by, in the form: (name, reverse).
+  *               for sort by feature id (e.g. natural sort), use an empty string for name
+  */
+class SortingSimpleFeatureIterator(features: CloseableIterator[SimpleFeature], sortBy: Seq[(String, Boolean)])
+    extends CloseableIterator[SimpleFeature] {
+
+  import SortingSimpleFeatureIterator.{sortInMemory, sortWithSpillover}
+
+  private val closed = new AtomicBoolean(false)
+
+  private lazy val sorted: CloseableIterator[SimpleFeature] = {
+    if (closed.get || !features.hasNext) { features } else {
+      val head = features.next()
+      if (closed.get || !features.hasNext) { CloseableIterator.single(head, features.close()) } else {
+        val ordering = SimpleFeatureOrdering(head.getFeatureType, sortBy)
+        QueryProperties.SortMemoryThreshold.toBytes match {
+          case None => sortInMemory(head, features, ordering, closed)
+          case Some(threshold) => sortWithSpillover(head, features, ordering, closed, threshold)
+        }
+      }
+    }
+  }
+
+  override def hasNext: Boolean = sorted.hasNext
+
+  override def next(): SimpleFeature = sorted.next()
+
+  override def close(): Unit = {
+    if (closed.compareAndSet(false, true)) {
+      sorted.close() // also closes original features iterator
+    }
+  }
+}
+
+object SortingSimpleFeatureIterator extends LazyLogging {
+
+  import scala.collection.JavaConverters._
+
+  /**
+   * Sorts the iterator in memory
+   *
+   * @param head head of iterator
+   * @param tail tail of iterator
+   * @param ordering ordering
+   * @param closed checked to determine if we should stop sorting and return early
+   * @return
+   */
+  private def sortInMemory(
+      head: SimpleFeature,
+      tail: CloseableIterator[SimpleFeature],
+      ordering: Ordering[SimpleFeature],
+      closed: AtomicBoolean): CloseableIterator[SimpleFeature] = {
+    // use ArrayList for sort-in-place of the underlying array
+    val list = new java.util.ArrayList[SimpleFeature](100)
+    list.add(head)
+    // note: we already ensured tail.hasNext == true before calling this method
+    do { list.add(tail.next()) } while (tail.hasNext && !closed.get)
+    if (closed.get) {
```

Comment:
Ok, in what case does the list get sorted *right* it is closed?  

Also, I think a newline above this line may increase readability.  While scanning the code, I was looking for {} associated to the 'while'.  (It took a split second to see the 'do'.)

Related to the do-while, I wonder if it is better as while (even though we have already done the first checks previously).

Reply:
> Ok, in what case does the list get sorted right it is closed?
can you clarify what you mean?
I can fix the formatting and get rid of the `do`

Reply:
Ah, I think I meant to ask ....

"In what cases does the list get sorted and then immediately closed?"  Aka, is this check necessary?  (If it is a rare case, we can leave it.)

Reply:
this would be to cover the query timeout kill. the while loop is checking the `closed` val, so if the query gets killed before it completes, we'll exit the loop and then this check lets us just return immediately without sorting.